src/dmd/traits.d

```diff
@@ -1332,7 +1332,17 @@ extern (C++) Expression semanticTraits(TraitsExp e, Scope* sc)
             return 0;
         }
 
-        ScopeDsymbol._foreach(sc, sds.members, &pushIdentsDg);
+        __gshared int counter;
+        if (counter > 1000)
+        {
+            e.error("Maximal recursion depth hit");
+        }
+        else
+        {
+            counter++;
+            ScopeDsymbol._foreach(sc, sds.members, &pushIdentsDg);
+        }
```

Comment:
decrease counter here, otherwise you are counting static_foreach frequency and not just recustion-depth.